### PR TITLE
add setUsername functionality

### LINKED ISSUE #39: [contracts] Add function to claim a username
- Add a mapping of hashed usernames to booleans to prevent multiple users having the same username
- Add a function to attest to an updated graffiti value

### PR DESCRIPTION
**Pull request by yuriko627**
## Pull request checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [ ] Build (`yarn build`) was run locally and any changes were pushed
- [ ] Lint (`yarn lint --check`) has passed locally and any fixes were made for failures

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Closes #39 

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/core/contracts/UnirepSocial.sol
Comment (vimwitch) [inline packages/core/contracts/UnirepSocial.sol:32]: We don't need an array of usernames. We shouldn't ever need to iterate over them

File: packages/core/contracts/UnirepSocial.sol
Comment (vimwitch) [inline packages/core/contracts/UnirepSocial.sol:40]: Because this is the only state variable relating to usernames this should be named something related to usernames like `claimedUsernames` or even just `usernames`.

File: packages/core/contracts/UnirepSocial.sol
Comment (vimwitch) [inline packages/core/contracts/UnirepSocial.sol:89]: No need to accept a username in the constructor.

File: packages/core/contracts/UnirepSocial.sol
Comment (vimwitch) [inline packages/core/contracts/UnirepSocial.sol:339]: This internal logic is correct but we don't need the loop around it. You can modify `isClaimed` using `oldUsername` and `newUsername`. e.g.

```
isClaimed[oldUsername] = false;
isClaimed[newUsername] = true;
```

Generally we want to avoid iterating over arrays in contracts, especially when they can grow to large sizes. This function would eventually be impossible to execute because it would have so many iterations that it would exceed the block gas limit.

**Review by vimwitch — CHANGES_REQUESTED** [review]
The core logic is correct, just need to remove some of the stuff around it. This pr also needs tests for this function. Just testing to make sure a username can't be double registered and that the old username is freed.

I'm going to publish a new version of the unirep stuff to npm today so it should be easier to build/test.

File: packages/core/contracts/UnirepSocial.sol
Comment (vivianjeng) [inline packages/core/contracts/UnirepSocial.sol:105]: There is no `userNames` in `unirep`.
Just store the `userNames` in Unirep Social contract.

File: packages/core/contracts/UnirepSocial.sol
Comment (yuriko627) [reply packages/core/contracts/UnirepSocial.sol:40]: Ah now I understand how set of usernames are represented by this mapping.

**Comment by yuriko627** [comment]
I get this error for 'should fail if a username is double registered' test

'Error: cannot estimate gas; transaction may fail or may require manual gas limit'

I saw this https://github.com/ethereum/solidity/issues/13159, but I assumed it's not ok to add `allowUnlimitedContractSize: true`. Is there any other way to get around this error? Or this test should be written different way in the first place?

**Comment by vimwitch** [comment]
> Error: cannot estimate gas; transaction may fail or may require manual gas limi

This is a really generic error that occurs anytime the transaction might revert. In this case you need to send the transaction from the `admin` account to set a username, otherwise the `require(msg.sender == admin)` line will revert.

Try something like this:
```
const accounts = await ethers.getSigners()
await unirepSocialContract.connect(accounts[0]).setUsername(
  config.NUM_EPOCH_KEY_NONCE_PER_EPOCH,
  newUsername,
  newUsername
)
```

You also don't need the explicit assertions for tx success. You can replace 
```
const tx = await unirepSocialContract.setUsername(
  config.NUM_EPOCH_KEY_NONCE_PER_EPOCH,
  newUsername,
  newUsername
)
const receipt = await tx.wait()
expect(receipt.status).equal(0)
```
With
```
await unirepSocialContract.setUsername(
  config.NUM_EPOCH_KEY_NONCE_PER_EPOCH,
  newUsername,
  newUsername
).then(t => t.wait())
```

The `wait` promise will throw an error if the transaction is not successful.

**Comment by vimwitch** [comment]
Ah I totally forgot something for this PR. We need to actually attest to the epoch key to give the key the username! This functionality is blocked by #30 because you need the attestation function. You'll use the [`submitAttestation`](https://github.com/Unirep/Unirep/blob/main/packages/contracts/contracts/Unirep.sol#L333) function in the Unirep contract. [Here](https://github.com/Unirep/Unirep-Social/blob/unirep-update/packages/core/contracts/UnirepSocial.sol#L182) is an example use. You'll want positive rep and negative rep to be 0, and the graffiti should be the new username.

You can rebase this branch on top of `unirep-update` now if you like, or you can wait for the pr to be merged.

**Comment by yuriko627** [comment]
I added `submitAttestation` part in the `setUsername` function, but `submitAttestation` function call failed to execute in the test.

File: .node-version
Comment (vimwitch) [inline .node-version:1]: These files should be removed. If they're created automatically by your environment or IDE you can add them to the `.gitignore` file.

File: package.json
Comment (vimwitch) [inline package.json:13]: These should not be added to the root level package file.

File: packages/core/contracts/UnirepSocial.sol
Comment (vimwitch) [inline packages/core/contracts/UnirepSocial.sol:284]: We have more usable getter functions for these values now. This can be replaced with `uint attestingFee = unirep.attestingFee()`.

File: packages/core/contracts/UnirepSocial.sol
Comment (vimwitch) [inline packages/core/contracts/UnirepSocial.sol:298]: `posRep` and `negRep` should both be 0 here.

File: packages/core/test/SignUp.ts
Comment (vimwitch) [inline packages/core/test/SignUp.ts:21]: These shouldn't be defined at the top level. Generally we want to avoid sharing variables or state between tests. This is an anti-pattern we have in a lot of our test files that we're slowly removing.

File: packages/core/test/SignUp.ts
Comment (vimwitch) [inline packages/core/test/SignUp.ts:90]: The first argument to this function is the epoch key that is being attested to. You should use a random value for this, or even just a hardcoded value like `0x2140`. There's a function in `@unirep/crypto` called `genRandomSalt()`. You should able to use that to get safe random epoch keys.

Definitely don't this config value here, it doesn't really make sense.

File: packages/core/test/SignUp.ts
Comment (vimwitch) [inline packages/core/test/SignUp.ts:105]: This test should have a failure condition. It should be completely separate from the last test (don't use the same usernames).

The flow should be:
```
Register username u1 to epoch key e1
Try to register username u1 to epoch key e2 (expecting a failure)
```

File: packages/core/test/SignUp.ts
Comment (vimwitch) [inline packages/core/test/SignUp.ts:116]: This test can be removed. It shouldn't be possible to make this case fail. If you want to keep it in user a fresh username for testing, don't use one that is shared with other tests.

Generally we don't want tests depending on each other because a single failure causes a cascade, and it introduces hard to diagnose bugs.

**Review by vimwitch — CHANGES_REQUESTED** [review]

File: packages/core/test/SignUp.ts
Comment (yuriko627) [reply packages/core/test/SignUp.ts:90]: Is it ok to set epochKey's type as `epochKey: BigNumberish | BigInt` ? cuz currently `BigNumberish` does not include bigint/BigInt. ref: https://github.com/ethers-io/ethers.js/discussions/1471 
You get this error saying `Argument of type 'BigInt' is not assignable to parameter of type 'BigNumberish'.
  Type 'BigInt' is missing the following properties from type 'BigNumber': _hex, _isBigNumber, fromTwos, toTwos, and 24 more.`  with `genRandomSalt()` which returns BigInt

File: packages/core/test/SignUp.ts
Comment (vimwitch) [inline packages/core/test/SignUp.ts:144]: Other test cases:

- Should fail if called by non-admin
- Should mark old username as not claimed
- Should emit attestation event with graffiti matching the new username
  - [This](https://github.com/Unirep/Unirep/blob/main/packages/contracts/contracts/interfaces/IUnirep.sol#L22) is what the attestation event looks like
  - [This](https://hardhat.org/hardhat-runner/plugins/nomicfoundation-hardhat-chai-matchers#usage) is how you can test that a contract call emitted an event
  - Would be great if you could check the contents of the event to make sure it matches the username that was set. I'm not sure how to do it in the chai matchers though.
  - [This](https://github.com/Unirep/Unirep/blob/main/packages/core/src/Synchronizer.ts#L478) is an example decoding an attestation event, maybe it will help

File: yarn.lock
Comment (vimwitch) [inline yarn.lock:2]: This needs to be regenerated after removing the deps from the root package. Just run `yarn` in the root directory and commit the changes.

**Review by vimwitch — CHANGES_REQUESTED** [review]

File: packages/core/.node-version
Comment (vimwitch) [inline packages/core/.node-version:1]: This file still needs to be removed

File: packages/core/test/SignUp.ts
Comment (vimwitch) [inline packages/core/test/SignUp.ts:183]: Here you should claim the username before changing it. Otherwise the default value of the old username is 0 and we’re not testing anything

File: packages/core/contracts/UnirepSocial.sol
Comment (vivianjeng) [inline packages/core/contracts/UnirepSocial.sol:300]: The `username` should be hashed by a `Poseidon` hash function
so in the reputation proof the user can prove the username is the pre-image of the graffiti.

Comment (vivianjeng) [reply packages/core/contracts/UnirepSocial.sol:300]: I am not sure whether it will be included in the PR because the `Poseidon` library should be linked as well.

Comment (vimwitch) [reply packages/core/contracts/UnirepSocial.sol:300]: I was planning to hash it off chain so we don’t have to worry about that.

**Comment by vimwitch** [comment]
@yuriko627 the files just need to be linted then this lgtm

File: packages/core/test/SignUp.ts
Comment (vimwitch) [inline packages/core/test/SignUp.ts:183]: This is just a note, instead of using like `tx1/tx2` variables you can put this logic inside a block like this:

```js
...
const newUsername4 = genRandomSalt().valueOf()
const accounts = await ethers.getSigners()
{
  const tx = await unirepSocialContract....
  const receipt = await tx.wait()
  expect(receipt.status)...
}
{
  const tx = await unirepSocialContract....
  const receipt = await tx.wait()
  expect(receipt.status)...
}
```

This bracket blocks can exist anywhere in the code, not just around `if` or `for` or functions or other flow control structures. Then the variables are scoped to that block and gone after.

**Review by vimwitch — APPROVED** [review]

File: packages/core/contracts/UnirepSocial.sol
Comment (vivianjeng) [inline packages/core/contracts/UnirepSocial.sol:280]: I am not sure whether `epochKey` should be verified on-chain so that it will not be easily changed by other users (if we don't want an `admin` to control)

**Review by vivianjeng — APPROVED** [review]
I think `epochKey` verification should be either in the backend or in the contract.
But it can be in another PR.
This PR LGTM.

**Pull request merged**
